src/BoundsInference.cpp

```diff
@@ -159,18 +158,17 @@ class BoundsInference : public IRMutator {
                 // figure out what those dimensions are, and just have all
                 // stages but the last use the bounds for the last stage.
                 vector<bool> always_pure_dims(func.args().size(), true);
-                const std::vector<UpdateDefinition> &updates = func.updates();
-                for (size_t i = 0; i < updates.size(); i++) {
+                for (UpdateDefinition i : func.updates()) {
                     for (size_t j = 0; j < always_pure_dims.size(); j++) {
-                        const Variable *v = updates[i].args[j].as<Variable>();
+                        const Variable *v = i.args[j].as<Variable>();
                         if (!v || v->name != func.args()[j]) {
                             always_pure_dims[j] = false;
                         }
                     }
                 }
 
-                if (stage < (int)func.updates().size()) {
-                    size_t stages = func.updates().size();
+                if (stage < func.num_update_definitions()) {
```

Comment:
This surfaces an unexamined API decision. Generally exposing a collection via a method to index the collection and another num_ method is not very C++ like. E.g. one cannot use a ranged for loop to iterate all the update definitions. The right way to do this is for updates() to return an iterable collection with a size method and an operator[]. (The collection does not have to be concretely exported as e.g. a vector of course.)